#! /bin/sh
patch -p1 -l -f $* < $0
exit $?

From 0a9cc0bd2e9c07682b07f517736a6d8b20c8a97f Mon Sep 17 00:00:00 2001
From: Kai Uwe Broulik <kde@privat.broulik.de>
Date: Mon, 9 Sep 2024 11:43:43 +0200
Subject: [PATCH] kstyle: Support registering QMainWindow in tools area
 manager

When calling QWindowSystemInterface::handleThemeChange Qt will
unpolish and re-polish *the window* but none of its child widgets.

When the window is unpolished, all its toolbars are unregistered.
Then the window is polished and we don't re-register the toolbars,
breaking the tools area.

(cherry picked from commit c6fa8b4e542d38071efcf7c0f4357440abfaf7ef)
---
 kstyle/breezetoolsareamanager.cpp | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/kstyle/breezetoolsareamanager.cpp b/kstyle/breezetoolsareamanager.cpp
index 504c4424..27cf81e1 100644
--- a/kstyle/breezetoolsareamanager.cpp
+++ b/kstyle/breezetoolsareamanager.cpp
@@ -257,8 +257,18 @@ void ToolsAreaManager::registerWidget(QWidget *widget)
     Q_ASSERT(widget);
     auto ptr = QPointer<QWidget>(widget);

+    QPointer<QMainWindow> mainWindow = qobject_cast<QMainWindow *>(ptr);
+
+    if (mainWindow && mainWindow == mainWindow->window()) {
+        const auto toolBars = mainWindow->findChildren<QToolBar *>(QString(), Qt::FindDirectChildrenOnly);
+        for (auto *toolBar : toolBars) {
+            tryRegisterToolBar(mainWindow, toolBar);
+        }
+        return;
+    }
+
     auto parent = ptr;
-    QPointer<QMainWindow> mainWindow = nullptr;
+
     while (parent != nullptr) {
         if (qobject_cast<QMdiArea *>(parent) || qobject_cast<QDockWidget *>(parent)) {
             break;
--
2.49.0
